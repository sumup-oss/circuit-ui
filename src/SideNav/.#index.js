ilyanoskov@Ilyas-MacBook-Air.local.27464