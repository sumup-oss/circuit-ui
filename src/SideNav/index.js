import React from 'react';
import withStyles from '../../util/withStyles';
import styles from './index.scss';

const SideNavComponent = props => {
  const status = props.expanded ? 'expanded' : 'collapsed';
  const childrenWithProps = React.Children.map(props.children, child =>
    React.cloneElement(child, {
      expanded: props.expanded
    })
  );

  return (
    <div className={`layout-sidenav layout-sidenav--${status}`}>
      <ul className={`sidenav-list sidenav-list--${status}`}>
        {childrenWithProps}
      </ul>
    </div>
  );
};

const MenuItemComponent = props => {
  const itemLinkStatus = props.isActive ? 'sidenav-item-link--active' : '';
  const itemTextStatus = props.expanded ? 'expanded' : 'collapsed';
  let click = () => {};
  if (props.onClick) click = props.onClick;

  return (
    <li className="sidenav-item " onClick={click}>
      <a className={`sidenav-item-link ${itemLinkStatus}`}>
        <span className="sidenav-icon--container">
          {React.cloneElement(props.icon, {
            className: 'icon sidenav-icon',
            viewBox: '0 0 24 24',
            width: '20',
            height: '20'
          })}
        </span>
        <span
          className={`sidenav-item-text sidenav-item-text--${itemTextStatus}`}
        >
          {props.children}
        </span>
      </a>
    </li>
  );
};

const SideNav = withStyles(styles)(SideNavComponent);
const MenuItem = withStyles(styles)(MenuItemComponent);

export { SideNav, MenuItem };
