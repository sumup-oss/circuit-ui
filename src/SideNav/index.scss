@import './../shared';

$max-width: 960px;
$border-box-sizing: false;

//-------------------------------------
// Overrides for Neat
//-------------------------------------

$column: golden-ratio(1em, 3) !default;
$gutter: golden-ratio(1em, 3) !default;
$fg-column: $column;
$fg-gutter: $gutter;

$topnav-height: 60px;
$topnav-height-mobile: 48px;
$expanded-side-nav-width: 225px;
$collapsed-side-nav-width: 60px;

body {
  background-color: $color-off;
  font-size: $font-size-m;
}

.layout-sidenav {
  background-color: $color-light;
  box-shadow: 1px 0 5px $color-medium;
  height: 100%;
  left: -$expanded-side-nav-width;
  position: fixed;
  top: $topnav-height-mobile;
  transition: .3s ease-in;
  z-index: 40;

  &.layout-sidenav--expanded {
    left: 0;
    transition: .3s ease-in;
    width: $expanded-side-nav-width;
  }

  @include media($big-screens) {
    box-shadow: none;
    left: 0;
    top: $topnav-height;
    transition: .3s ease-in;
    width: $collapsed-side-nav-width;
  }
}

.wrapper {
  width: 100%;
}

.sidenav-list {
  margin-top: 20px;
  padding: 0;
  text-transform: uppercase;
}

.sidenav-item {
  @include display(flex);

  margin: 10px auto 10px 0;

  &#{&}--fade-in {
    @include animation(fadeInDown .2s ease-in-out);
  }
}

.sidenav-item-text {
  display: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;

  &.sidenav-item-text--expanded {
    display: inline-block;
  }
}

.sidenav-item-link {
  @include type-m;
  $left-border-width: $spacing-xxs;

  @include display(flex);
  @include align-items(center);

  border-left: $left-border-width solid transparent;
  color: $color-dark;
  height: 32px;
  max-width: 100%;
  padding: 7px 12px 0 14px;

  &.sidenav-item-link--active {
    border-left: $left-border-width solid $color-brand;
    font-weight: 600;
  }

  &.sidenav-item-link--disabled {
    color: $color-medium;
    cursor: default;
    transform-style: preserve-3d;

    &:hover {
      color: $color-medium;
    }
  }

  &:hover {
    color: $color-dark;
  }
}

.sidenav-icon-container {
  margin-right: $spacing-l;
}

.sidenav-icon {
  fill: $color-medium;
  margin-right: 20px;

  use > svg {
    fill: $color-medium;
  }
}
