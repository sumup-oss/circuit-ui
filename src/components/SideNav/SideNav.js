/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/** @jsx jsx */

import { jsx, css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import {
  Drawer,
  List,
  ListItem,
  ListItemIcon,
  ListItemText
} from './components';
import { childrenPropType } from '../../util/shared-prop-types';

import { disableVisually } from '../../styles/style-helpers';

export const COLLAPSED_NAV_WIDTH = 64;
export const FULL_NAV_WIDTH = 270;

export const MODE = {
  FULL: 'full',
  CONDENSED: 'condensed',
  HIDDEN: 'hidden'
};

const getDrawerContentStyles = ({ mode }) => css`
  width: ${mode === MODE.CONDENSED ? COLLAPSED_NAV_WIDTH : FULL_NAV_WIDTH}px;
  overflow: hidden;
  background-color: #212933;
  color: white;
  transition: width 225ms cubic-bezier(0.4, 0, 0.6, 1) 0ms;
  height: ${window.innerHeight}px;
`;

const NavListItem = styled(({ icon, children, ...rest }) => (
  <ListItem {...rest} component="a">
    {!!icon && <ListItemIcon>{icon}</ListItemIcon>}
    <ListItemText>{children}</ListItemText>
  </ListItem>
))`
  color: white;

  &:hover:not([disabled]) {
    background-color: rgba(15, 19, 26, 0.5);
  }

  &[disabled] {
    ${disableVisually()};
  }
`;

const Container = styled(Drawer)`
  ${({ condensed }) => css`
    width: 0;
    width: ${condensed ? COLLAPSED_NAV_WIDTH : FULL_NAV_WIDTH}px;
    transition: width 0.3s ease-out;
    ${ListItemText} {
      ${condensed && 'display: none'};
    }
  `};
`;

const NavList = styled(List)`
  user-select: none;
  -ms-overflow-style: none;
  overflow-y: auto;
`;

const FooterWrapper = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1 0 auto;
  align-self: stretch;
  align-items: stretch;
  justify-content: flex-end;
`;

const SideNav = props => {
  const { variant, toggleMode, mode, header, footer, children } = props;
  const drawerContentStyles = getDrawerContentStyles(props);

  return (
    <Container
      condensed={mode === MODE.CONDENSED}
      onClose={() => toggleMode(MODE.HIDDEN)}
      open={mode !== MODE.HIDDEN}
      variant={variant}
      anchor="left"
      CardProps={{ additionalStyles: drawerContentStyles }}
    >
      <div
        css={css`
          flex-shrink: 0;
        `}
      >
        {header}
      </div>
      <NavList gutter={mode === MODE.FULL ? List.KILO : List.BYTE}>
        {children}
      </NavList>
      {!!footer && <FooterWrapper>{footer}</FooterWrapper>}
    </Container>
  );
};

SideNav.propTypes = {
  variant: PropTypes.oneOf(['permanent', 'temporary']).isRequired,
  toggleMode: PropTypes.func,
  mode: PropTypes.oneOf([MODE.FULL, MODE.CONDENSED, MODE.HIDDEN]).isRequired,
  header: PropTypes.node,
  footer: PropTypes.node,
  children: childrenPropType
};

SideNav.defaultProps = {
  toggleMode: () => {},
  header: null,
  footer: null,
  children: null
};

SideNav.Item = NavListItem;
SideNav.MODE = MODE;
SideNav.COLLAPSED_NAV_WIDTH = COLLAPSED_NAV_WIDTH;
SideNav.FULL_NAV_WIDTH = FULL_NAV_WIDTH;

export default SideNav;
