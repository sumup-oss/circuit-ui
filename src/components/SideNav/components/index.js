/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as Fade } from './Fade';
export { default as Backdrop } from './Backdrop';
export { default as RootRef } from './RootRef';
export { default as Modal } from './Modal';
export { default as Drawer } from './Drawer';
export { default as List } from './List';
export { default as ListItem } from './ListItem';
export { default as ListItemText } from './ListItemText';
export { default as ListItemIcon } from './ListItemIcon';
export { default as Slide } from './Slide';
export { default as Menu } from './Menu';
