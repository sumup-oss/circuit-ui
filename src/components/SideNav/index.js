export { default } from './SideNav';

export * from './SideNav';

export { Drawer } from './components';

export { Menu } from './components';
