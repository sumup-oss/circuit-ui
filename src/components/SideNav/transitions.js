/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable no-param-reassign */
/* eslint-disable no-restricted-globals */

import warning from 'tiny-warning';

// Follow https://material.google.com/motion/duration-easing.html#duration-easing-natural-easing-curves
// to learn the context in which each easing should be used.
const easing = {
  // This is the most common easing curve.
  easeInOut: 'cubic-bezier(0.4, 0, 0.2, 1)',
  // Objects enter the screen at full velocity from off-screen and
  // slowly decelerate to a resting point.
  easeOut: 'cubic-bezier(0.0, 0, 0.2, 1)',
  // Objects leave the screen at full velocity. They do not decelerate when off-screen.
  easeIn: 'cubic-bezier(0.4, 0, 1, 1)',
  // The sharp curve is used by objects that may return to the screen at any time.
  sharp: 'cubic-bezier(0.4, 0, 0.6, 1)'
};

// Follow https://material.io/guidelines/motion/duration-easing.html#duration-easing-common-durations
// to learn when use what timing
const duration = {
  shortest: 150,
  shorter: 200,
  short: 250,
  // most basic recommended timing
  standard: 300,
  // this is to be used in complex animations
  complex: 375,
  // recommended when something is entering screen
  enteringScreen: 225,
  // recommended when something is leaving screen
  leavingScreen: 195
};

export const formatMs = milliseconds => `${Math.round(milliseconds)}ms`;
export const isString = value => typeof value === 'string';
export const isNumber = value => !isNaN(parseFloat(value));

export const reflow = node => node.scrollTop;

export function getTransitionProps(props, options) {
  const { timeout, style = {} } = props;

  return {
    duration:
      style.transitionDuration || typeof timeout === 'number'
        ? timeout
        : timeout[options.mode],
    delay: style.transitionDelay
  };
}

/**
 * @param {string|Array} props
 * @param {object} param
 * @param {string} param.prop
 * @param {number} param.duration
 * @param {string} param.easing
 * @param {number} param.delay
 */
export default {
  easing,
  duration,
  create(props = ['all'], options = {}) {
    const {
      duration: durationOption = duration.standard,
      easing: easingOption = easing.easeInOut,
      delay = 0,
      ...other
    } = options;
    warning(
      isString(props) || Array.isArray(props),
      'Circuit-UI: argument "props" must be a string or Array.'
    );
    warning(
      isNumber(durationOption) || isString(durationOption),
      `Circuit-UI: argument "duration"
      must be a number or a string but found ${durationOption}.`
    );
    warning(
      isString(easingOption),
      'Circuit-UI: argument "easing" must be a string.'
    );
    warning(
      isNumber(delay) || isString(delay),
      'Circuit-UI: argument "delay" must be a number or a string.'
    );
    warning(
      Object.keys(other).length === 0,
      `Circuit-UI: unrecognized argument(s) [${Object.keys(other).join(',')}]`
    );
    return (Array.isArray(props) ? props : [props])
      .map(
        animatedProp =>
          `${animatedProp} ${
            typeof durationOption === 'string'
              ? durationOption
              : formatMs(durationOption)
          } ${easingOption} ${
            typeof delay === 'string' ? delay : formatMs(delay)
          }`
      )
      .join(',');
  },
  getAutoHeightDuration(height) {
    if (!height) {
      return 0;
    }
    const constant = height / 36;
    // https://www.wolframalpha.com/input/?i=(4+%2B+15+*+(x+%2F+36+)+**+0.25+%2B+(x+%2F+36)+%2F+5)+*+10
    return Math.round((4 + 15 * constant ** 0.25 + constant / 5) * 10);
  }
};
