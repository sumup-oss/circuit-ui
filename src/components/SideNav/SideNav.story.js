import React, { Component } from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';

import { GROUPS } from '../../../.storybook/hierarchySeparators';
import withTests from '../../util/withTests';
import SideNav from './SideNav';
import SideNavItem from './components/SideNavItem';
import FavoriteIcon from './components/icons/favorite.svg';

/* eslint-disable */
const Link = ({ onClick, href, children }) => (
  <div
    onClick={e => {
      e.preventDefault();
      onClick(href);
    }}
  >
    {children}
  </div>
);
/* eslint-enabled */

const routes = [
  {
    label: 'Foo',
    href: 'https://foo.com'
  },
  {
    label: 'Bar',
    href: 'https://bar.com'
  },
  {
    label: 'Baz',
    href: 'https://baz.com'
  }
];

// Just an example. Usually your router of choice would take
// care of these things.
class NavState extends Component {
  constructor(props) {
    super(props);
    this.state = {
      route: 'https://foo.com'
    };

    this.handleChangeRoute = this.handleChangeRoute.bind(this);
  }

  handleChangeRoute(route) {
    action('changed route')(route);
    this.setState({ route });
  }

  render() {
    return this.props.children(this.state.route, this.handleChangeRoute);
  }
}

storiesOf(`${GROUPS.NAVIGATION}|SideNav`, module)
  .addDecorator(withTests('SideNav'))
  .add(
    'SideNav',
    withInfo()(() => (
      <NavState>
        {(route, handleChangeRoute) => (
          <SideNav>
            {routes.map(({ label, href }) => (
              <Link href={href} onClick={handleChangeRoute}>
                <SideNavItem isActive={route === href} icon={FavoriteIcon}>
                  {label}
                </SideNavItem>
              </Link>
            ))}
          </SideNav>
        )}
      </NavState>
    ))
  );
