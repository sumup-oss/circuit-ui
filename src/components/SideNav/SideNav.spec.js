import React from 'react';

import SideNav from './SideNav';

describe('SideNav', () => {
  describe('rendered styles', () => {
    it('should render with default styles', () => {
      const actual = create(
        <SideNav>
          <span>One</span>
        </SideNav>
      );
      expect(actual).toMatchSnapshot();
    });
  });

  describe('component logic', () => {
    it('should show a header when passsed', () => {
      // Component with header
      const wrapperWith = shallow(
        <SideNav header={<div data-test="header">Foobar</div>}>
          <span>Hello</span>
        </SideNav>
      );
      const actualWith = wrapperWith.find('div[data-test="header"]');
      expect(actualWith).not.toBeEmpty();

      // Component without header
      const wrapperWithout = shallow(
        <SideNav>
          <span data-test="header">Hello</span>
        </SideNav>
      );
      const actualWithout = wrapperWithout.find('NavHeader');

      expect(actualWithout).toBeEmpty();
    });

    describe('showing navigation items', () => {
      let wrapper;

      beforeEach(() => {
        wrapper = shallow(
          <SideNav>
            <span data-test="item">One</span>
            <span data-test="item">Two</span>
          </SideNav>
        );
      });

      it('should show its children in a list', () => {
        const actualChildren = wrapper.find('span[data-test="item"]');
        expect(actualChildren).toHaveLength(2);
      });

      // TODO: find a nicer way to test this?
      it('should show children in a list', () => {
        const children = wrapper.find('span[data-test="item"]');
        const listItems = children.map(child =>
          child
            .parent()
            .dive()
            .find('li')
        );
        listItems.forEach(item => {
          const actual = item.find('li');
          expect(actual).toHaveLength(1);
        });
      });
    });
  });

  describe('component accessibility', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(
        <SideNav>
          <span>One</span>
        </SideNav>
      );
      const actual = await axe(wrapper);
      expect(actual).toHaveNoViolations();
    });
  });
});
