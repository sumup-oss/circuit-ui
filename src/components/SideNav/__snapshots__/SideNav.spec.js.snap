// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav rendered styles should render with default styles 1`] = `
.circuit-4 {
  background-color: #212933;
  padding: 12px;
  width: 270px;
}

.circuit-2 {
  list-style: none;
}

.circuit-0:not(:last-of-type) {
  margin-bottom: 16px;
}

<nav
  className="circuit-4 circuit-5"
>
  <ul
    className="circuit-2 circuit-3"
  >
    <li
      className="circuit-0 circuit-1"
    >
      <span>
        One
      </span>
    </li>
  </ul>
</nav>
`;
